test/metabase/models/params/chain_filter_test.clj

```diff
@@ -69,6 +69,89 @@
+  50 -> end"
+  (let [big 50]
+    (merge-with merge
+                (reduce (fn [m [x y]] (assoc-in m [x y] [[x y]]))
+                        {}
+                        (for [x     (range (inc big))
+                              y     (range (inc big))
+                              :when (not= x y)]
+                          [x y]))
+                {:start (reduce (fn [m x] (assoc m x [[:start x]]))
+                                {}
+                                (range (inc big)))}
+                {big    {:end [[big :end]]}})))
+
+(def megagraph-single-path
+  "Similar to the megagraph above, this graph only has a single path through a hugely interconnected graph. A naive
+  graph traversal will run out of memory or take quite a long time to find the traversal:
+
+  [[:start 90] [90 200] [200 :end]]
+
+  There is only one path to end (from 200) and only one path to 200 from 90. If you take out the seen nodes this path
+  will not be found as the traversal advances through all of the 50 paths from start, all of the 50 paths from 1, all
+  of the 50 paths from 2, ..."
+  (merge-with merge
+              ;; every node is linked to every other node (1 ... 199)
+              (reduce (fn [m [x y]] (assoc-in m [x y] [[x y]]))
+                      {}
+                      (for [x     (range 200)
+                            y     (range 200)
+                            :when (not= x y)]
+                        [x y]))
+              {:start (reduce (fn [m x] (assoc m x [[:start x]]))
+                              {}
+                              (range 200))}
+              ;; only 90 reaches 200 and only 200 (big) reaches the end
+              {90  {200 [[90 200]]}
+               200 {:end [[200 :end]]}}))
+
+(deftest traverse-graph-test
+  (testing "If no need to join, returns immediately"
+    (is (nil? (#'chain-filter/traverse-graph {} :start :start 5))))
+  (testing "Finds a simple hop"
+    (let [graph {:start {:end [:start->end]}}]
+      (is (= [:start->end]
+             (#'chain-filter/traverse-graph graph :start :end 5))))
+    (testing "Finds over a few hops"
+      (let [graph {:start {:a [:start->a]}
+                   :a     {:b [:a->b]}
+                   :b     {:c [:b->c]}
+                   :c     {:end [:c->end]}}]
+        (is (= [:start->a :a->b :b->c :c->end]
+               (#'chain-filter/traverse-graph graph :start :end 5)))
+        (testing "But will not exceed the max depth"
+          (is (nil? (#'chain-filter/traverse-graph graph :start :end 2))))))
+    (testing "Can find a path in a dense and large graph"
+      (is (= [[:start 50] [50 :end]]
+             (#'chain-filter/traverse-graph megagraph :start :end 5)))
+      (is (= [[:start 90] [90 200] [200 :end]]
+             (#'chain-filter/traverse-graph megagraph-single-path :start :end 5))))
+    (testing "Returns nil if there is no path"
```

Comment:
nit: should prolly check that it respects the max-hops param, since putting that in is the thing that's gonna fix this bug either way

Reply:
line 129 checks this:
```clojure
   (testing "Finds over a few hops"
      (let [graph {:start {:a [:start->a]}
                   :a     {:b [:a->b]}
                   :b     {:c [:b->c]}
                   :c     {:end [:c->end]}}]
        (is (= [:start->a :a->b :b->c :c->end]
               (#'chain-filter/traverse-graph graph :start :end 5)))
        (testing "But will not exceed the max depth" ;; max depth set to 2 and it returns nil here
          (is (nil? (#'chain-filter/traverse-graph graph :start :end 2))))))
```

The thing that actually fixes this is the BFS change. Descending into traversals was the wrong strategy to find our way across big databases.